### ISSUE TITLE
Create SF Flows for WB survey project

### ISSUE THREAD

**Issue by samschlicht** [talentbeyondboundaries/tbbtalentv2#568]

**Comment by samschlicht** [comment]
Check the epic for more context, but here's the SF ask:
- When a job opportunity goes to CV review, all consenting candidates are automatically sent the baseline survey;
- unless: they've previously got to CV review stage in relation to another opportunity, in which case they remain part of that group.
- 6, 12 and 18 months after relocation of the successful candidate(s), follow-up surveys are sent to the same group;
- unless: any of the unsuccessful candidates for that opportunity are subsequently successful in another and thereby relocate — in which case they become part of that group, and cease to be sent the follow-up surveys for the previous opportunity.

As noted in the epic, we also may need something like [the Zapier automation for the old TBB impact evaluation surveys](https://zapier.com/webintent/edit-zap/154226625), which updates SF records, Slack and Google Sheets with respondent and workflow fail information.

Here are some early considerations in that regard:
- For the old TBB impact evaluation surveys, we ask candidates each time to fill in their name and email, and then use that to automate a record of their response on SF and elsewhere — once a month or so, we get an error because a candidate has changed their email or makes a typo on the survey. It's not a big deal to keep a spreadsheet of workflow fails and update manually in those instance, but if we can add a unique identifier to each Survey Monkey link sent out via SF, that would reduce the manual work both of our staff and the survey respondents.
- Same goes for identifying job opportunity 'groups' (since the candidates going for a particular job will ultimately be compared to one another, as well as to the general group) — quite easy to do semi-manually based on our data, but ideally would be automated via the SM link.
- Finally, we want candidates already on the old impact evaluation survey track to keep receiving those follow-ups, and to avoid them receiving the new ones (the latter should be easy, since they won't get the opportunity to give consent, which will always be a condition for sending the new ones)

The difficulty of this issue (and the accuracy of its estimate) will depend greatly on the limitations of SF Flows and Survey Monkey, etc, so I'll update with specifics as they come, and obviously return to the group at any key decision points.

**Comment by samschlicht** [comment]
Noting that I've checked in with Mariam re when intakes reliably happen — if the baseline survey send is triggered by CV review stage, we need to know that intake is always done at that point, otherwise we won't have candidate consent yet (since that's taken at intake). If so, we may need to add a delay or a retry to the trigger.

**Comment by samschlicht** [comment]
Also waiting on Sarah for Survey Monkey access.

**Comment by samschlicht** [comment]
The fun and games begin: 
- I need to get all the candidate opportunities attached to the employer opportunity that triggered the flow by moving to review stage, but they are linked by a lookup field, which can't be referenced in SF Flows.
- So what I've done is add a Parent Employer Opportunity ID field, only visible to system admins, to candidate opportunities. Now I can use it as a condition.

**Comment by samschlicht** [comment]
Made this additional section in Candidate Contact Records, visible to all but editable only by System Admins - the workflow will get and set these values.
<img width="998" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBMy9YQWc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--7376443dfa4d62f2b820e2e615044715d2cb3e5e/Screenshot%202024-01-18%20at%202.10.16%E2%80%AFpm.png" alt="Screenshot 2024-01-18 at 2.10.16 pm.png" />

**Comment by samschlicht** [comment]
I've decided to switch to Zapier for the automation — unfortunately SF Flows is just not up to this one. Every action has to be defined separately, I'm losing so much time switching between screens. It's also just so opaque, you have to test every tiny adjustment and as usual there's precious little doc.

**Comment by samschlicht** [comment]
Zapier looking good — great that I can combine Survey Monkey and Gmail in the same workflow. I think the next day I spend on this will be far more productive!

**Comment by samschlicht** [comment]
Figured out how to create personalised web links per respondent, allowing for proper results aggregation and tracking. [Slack post explaining to research group](https://refugeejobsmarket.slack.com/archives/C04MPLG345R/p1705880910368589).

Reminder to self that values for touchpoint are: [base, 6, 12, 18]

The Zap will populate these values for every survey it sends out.

**Comment by samschlicht** [comment]
For now I'm just putting generic comms in place. I've set up a new Google account <surveys@talentbeyondboundaries.org> for sending out the surveys, to which Sarah W has delegated access. The group will need to provide the following comms:
- Email for sending out base, 6, 12, 18 surveys incl. subject

So this is what the survey link ends up looking like in Zapier:
<img width="515" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBK1hqQWc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--4a5dea6682b8b43e32358808af5fad8d58e33f96/Screenshot%202024-01-22%20at%2011.19.25%E2%80%AFam.png" alt="Screenshot 2024-01-22 at 11.19.25 am.png" />

**Comment by samschlicht** [comment]
Baseline done for now — still needs:
- [ ] Complete candidate email
- [ ] Check that intake will be complete by the time CV review stage is triggered, and that staff will always use the CV review stage on a candidate opportunity record

**Comment by samschlicht** [comment]
**Sunsetting the old survey** 
- We need candidates who already responded to the old one to keep going on that track, but candidates who haven't responded to the old one to go on the new track and not be sent the old one. 
- The old one runs for 5 years after relocation, so we can't just stop it.
- The new baseline survey is not at issue, because a candidate who's previously accepted a job offer (and thereby is already on the old survey track) won't be put into CV review stage relating to another opportunity.
- It's also the case that any candidate who's already relocated or due to relocate for a job opportunity won't do another intake, and so won't be asked for monitoring and evaluation consent, thus would fail that condition for sending the new surveys.
- So, without any additional checks, it's already the case that no one on the old survey track will be sent the new one.

However, we will need a check to stop the old surveys being sent to a candidate on the new survey track:
- [x] Before sending an old survey, check that no new survey has ever been sent to them

**Comment by samschlicht** [comment]
Just thought I'd note for posterity the reasoning for my tool choices:

- SF Flows for the workflow that triggers survey send and updates the relevant records: being a native tool, the triggers and conditions are a bit more elaborate than Zapier — this enables me to allow for the different ways that someone might be marked as relocated (the process can vary a bit, and sometimes there's human error), for example.
- Zapier for the email survey send and response logging: SF Flows' email tool doesn't allow merge fields in hyperlinks, which I'm using to inject URL params for response grouping and analysis in Survey Monkey. In SF you also have to define the triggering workflow, email send action and email template separately, so the whole process is life-sappingly slow. I need Zapier for the response logging since the trigger for that is in Survey Monkey and Zapier enables combination of different tools.

In the end I actually found SF Flows very powerful for stuff happening within SF, but as soon you need to get beyond SF, it falters and you need Zapier.

**Comment by samschlicht** [comment]
I'm softening on SF Flows a little, but the email issue still applies, so I have to finish the workflow off in Zapier (triggering the actual email send).

[Here's the 6-month post-relocation SF Flow](https://talentbeyondboundaries.lightning.force.com/builder_platform_interaction/flowBuilder.app?flowId=301Uu00000270ADIAY), which does the following:
- Triggered by a candidate opportunity stage change to 'relocated' OR 'settled' OR 'won' AND relocated_date NOT NULL, then executes 6 months after the relocated date)
- Finds the sibling candidate opps (children of the same parent job opp, including the triggering record) and iterates through them, checking the '6-month survey sent' box when the following conditions apply:
- Candidate contact 'monitoring & evaluation consent' box is checked
- The parent employer opportunity of the triggering child candidate opportunity has the same ID as the candidate contact 'Employer opportunity group' (one candidate could be unsuccessful in multiple opps, but we want them to stay in their first group to avoid duplicates and retain a stable control group)
- OR - this candidate is the successful one (they got the job), in which case we _do_ want them to switch to this new group, since they're the candidate against which the control group will be compared. In this instance, the workflow takes an extra step and amends that candidate contact's 'employer opportunity group' to reflect the new one.

**Comment by samschlicht** [comment]
Added new survey track check to the old survey automations — this checks if any of the new surveys have been sent to this candidate contact. If they have, it doesn't send the old one.

<img width="740" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBN2pxQWc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--3478ca4b87762b3b6c27a9b51aab1a5b37f7e193/Screenshot%202024-01-23%20at%2012.41.40%E2%80%AFpm.png" alt="Screenshot 2024-01-23 at 12.41.40 pm.png" />

**Comment by samschlicht** [comment]
Finally: summarise work done and action items left, correct comments and check for any missed points

**Comment by samschlicht** [comment]
Okay, in terms of our team's work, this is at review stage, so here's a summary of where it's at:
(Scroll to the top comment in this issue to see my summary of the technical ask.)

Workflows:
(Zapier — the team can log in using tech@talentbeyondboundaries.org with pass: syd8B_zU_Q*3g4*)

- [Baseline send](https://zapier.com/webintent/edit-zap/222251413)
- [6 months post-relocation send](https://zapier.com/webintent/edit-zap/222712452)
- [12 months post-relocation send](https://zapier.com/webintent/edit-zap/222887085)
- [18 months post-relocation send](https://zapier.com/webintent/edit-zap/222887439)
- [Survey response box checking - all touchpoints](https://zapier.com/webintent/edit-zap/222705407)

(Salesforce — these verify the conditions and then trigger the Zapier automation to actually send the survey)
- [6 months post-relocation](https://talentbeyondboundaries.lightning.force.com/builder_platform_interaction/flowBuilder.app?flowId=301Uu00000270ADIAY)
- [12 months post-relocation](https://talentbeyondboundaries.lightning.force.com/builder_platform_interaction/flowBuilder.app?flowId=301Uu00000278pLIAQ)
- [18 months post-relocation](https://talentbeyondboundaries.lightning.force.com/builder_platform_interaction/flowBuilder.app?flowId=301Uu0000027B2PIAU)
- [e.g. of adjustment to old survey workflow](https://talentbeyondboundaries.lightning.force.com/builder_platform_interaction/flowBuilder.app?flowId=301Uu00000279AHIAY)

There are a number of action items left on this, but they are all about the implementation and will probably not greatly affect the underlying automations (I hope 🤞) — I'm putting this into review to get the team's eyes, and the rest will move forward as #579.
